spec/API_specification/array_api/array_object.py

```diff
@@ -473,37 +500,57 @@ def __gt__(self: array, other: Union[int, float, array], /) -> array:
            Element-wise results must equal the results returned by the equivalent element-wise function :func:`~array_api.greater`.
         """
 
-    def __index__(self: array, /) -> int:
+    def __index__(self: array, /) -> Union[int, bool]:
         """
-        Converts a zero-dimensional integer array to a Python ``int`` object.
+        Converts a zero-dimensional array to a Python object used in indexing.
 
         .. note::
            This method is called to implement `operator.index() <https://docs.python.org/3/reference/datamodel.html#object.__index__>`_. See also `PEP 357 <https://www.python.org/dev/peps/pep-0357/>`_.
 
         Parameters
         ----------
         self: array
-            zero-dimensional array instance. Must have an integer data type.
+            zero-dimensional array instance. Should have an integer or boolean data type. If ``self`` has a floating-point or complex data type, the function must raise a ``TypeError``.
 
         Returns
         -------
-        out: int
-            a Python ``int`` object representing the single element of the array instance.
+        out: Union[int, bool]
+            a Python object representing the single element of the array instance. If ``self`` has an integer data type, ``out`` must be a Python ``int`` object. If ``self`` has a boolean data type, ``out`` must be a Python ``bool`` object.
         """
 
     def __int__(self: array, /) -> int:
         """
-        Converts a zero-dimensional integer array to a Python ``int`` object.
+        Converts a zero-dimensional array to a Python ``int`` object.
+
+        .. note::
+           Casting floating-point values outside the representable bounds of Python's non-arbitary integer type is not specified and is implementation-dependent.
```

Comment:
I'm unclear why this is unspecified. Python's `int` allows any integer value of any size. There is an internal split between machine ints and big ints, but this not exposed to the user in the Python API in any way.

Reply:
Maybe just say `int64` (or `Py_ssize_t`, even if it is probably always 64bit), since on the Python side there is no non-arbitrary integer anyway?

Reply:
I'm just unclear why something like `int(asarray(1.7976931348623157e+308))` would ever not be able to work.

Is the point rather that anything greater than 1e15 (or thereabouts) might is not be fully defined? NumPy gives `179769313486231570814527423731704356798070567525844996598917476803157260780028538760589558632766878171540458953514382464234321326889464182768467546703537516986049910576551282076245490090389328944075868508455133942304583236903222948165808559332123348274797826204144723168738177180919299881250404026184124858368` for the above but I suppose it could also give some other value of the same magnitude that agrees in the first 15 digits (really should be the first 53 bits, not 15 decimal digits).

Reply:
Right, going from float -> int, I think it is always OK to do `int(<double>float32)`.  And for that Python provides the C-API.  So presumably for all dtypes that are explicitly listed in the standard there is really no reason why it would not be easy to provide the functionality.

Reply:
For context, I wrote this note because passing floating numbers larger than `sys.maxsize` was inconsistent, e.g.

```python
>>> sys.maxsize
9223372036854775807
>>> int(sys.maxsize + .1)
9223372036854775808  # expected 9223372036854775807
```

I see now that this isn't specifically because the numbers were larger than `sys.maxsize`, but because `float64` obviously isn't precise enough for all int64s, let alone big ints :man_facepalming:

So yeah, think I can just remove this note.